File: tests/cuda/python/test_indexing.py
"""Indexing and sorting CUDA binding tests."""

import sys
import os
import pytest

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
    import numpy as np
except ImportError:
    pytest.skip("Insight or NumPy not available", allow_module_level=True)

try:
    ins.init()
    if ins.device_count() == 0:
        pytest.skip("No GPU available", allow_module_level=True)
except Exception:
    pytest.skip("GPU backend not available", allow_module_level=True)

GPU = ins.GPUPlace(0)
CPU = ins.CPUPlace()


def to_gpu(np_arr, dtype=None):
    if dtype is not None:
        np_arr = np_arr.astype(dtype)
    return ins.from_numpy(np_arr).to(GPU)


def to_numpy(gpu_arr):
    return gpu_arr.to(CPU).numpy()


class TestIndexingCUDA:
    def test_take_1d(self):
        a = to_gpu(np.array([10, 20, 30, 40, 50], dtype=np.float64))
        idx = to_gpu(np.array([0, 2, 4], dtype=np.int64))
        result = ins.take(a, idx)
        np.testing.assert_allclose(to_numpy(result), np.array([10, 30, 50]))

    def test_nonzero(self):
        a = to_gpu(np.array([0, 1, 0, 3, 0], dtype=np.float64))
        result = ins.nonzero(a)
        assert result.numel > 0

    def test_flatnonzero(self):
        a = to_gpu(np.array([[0, 1], [2, 0]], dtype=np.float64))
        result = ins.flatnonzero(a)
        assert result.numel == 2

    def test_unique_1d(self):
        a = to_gpu(np.array([3, 1, 2, 1, 3, 2], dtype=np.float64))
        result = ins.unique(a)
        assert hasattr(result, "unique")
        vals = sorted(to_numpy(result.unique).tolist())
        assert vals == [1.0, 2.0, 3.0]

    def test_unique_sorted(self):
        a = to_gpu(np.array([5, 3, 1, 3, 5], dtype=np.float64))
        result = ins.unique(a)
        data = to_numpy(result.unique)
        assert np.all(np.diff(data) >= 0)

    def test_argsort_1d(self):
        a = to_gpu(np.array([3, 1, 2], dtype=np.float64))
        result = ins.argsort(a)
        expected = np.argsort(np.array([3, 1, 2]))
        np.testing.assert_array_equal(to_numpy(result), expected)

    def test_sort_1d(self):
        a = to_gpu(np.array([3, 1, 4, 1, 5], dtype=np.float64))
        result = ins.sort(a)
        expected = np.sort(np.array([3, 1, 4, 1, 5]))
        np.testing.assert_allclose(to_numpy(result), expected)

    def test_topk(self):
        a = to_gpu(np.array([1, 5, 3, 2, 4], dtype=np.float64))
        values, indices = ins.topk(a, 3)
        data = sorted(to_numpy(values).tolist(), reverse=True)
        assert data == [5.0, 4.0, 3.0]

    def test_gather(self):
        a = to_gpu(np.array([[10, 20], [30, 40]], dtype=np.float64))
        idx = to_gpu(np.array([[0, 1], [1, 0]], dtype=np.int64))
        result = ins.gather(a, idx, 0)
        assert result.numel == 4

    def test_scatter(self):
        a = to_gpu(np.zeros([4], dtype=np.float64))
        idx = to_gpu(np.array([0, 2], dtype=np.int64))
        src = to_gpu(np.array([10, 20], dtype=np.float64))
        result = ins.scatter(a, idx, src, 0)
        assert result.numel == 4

    def test_scatter_add(self):
        a = to_gpu(np.zeros([4], dtype=np.float64))
        idx = to_gpu(np.array([0, 0, 2], dtype=np.int64))
        src = to_gpu(np.array([1, 2, 3], dtype=np.float64))
        result = ins.scatter_add(a, idx, src, 0)
        assert result.numel == 4
        data = to_numpy(result)
        assert data[0] == pytest.approx(3.0)

    def test_searchsorted(self):
        a = to_gpu(np.array([1, 3, 5, 7], dtype=np.float64))
        v = to_gpu(np.array([0, 2, 5, 8], dtype=np.float64))
        result = ins.searchsorted(a, v)
        expected = np.searchsorted(np.array([1, 3, 5, 7]), np.array([0, 2, 5, 8]))
        np.testing.assert_array_equal(to_numpy(result), expected)

    def test_interp(self):
        xp = to_gpu(np.array([0, 1, 2, 3], dtype=np.float64))
        fp = to_gpu(np.array([0, 2, 4, 6], dtype=np.float64))
        x = to_gpu(np.array([0.5, 1.5, 2.5], dtype=np.float64))
        result = ins.interp(x, xp, fp)
        expected = np.interp([0.5, 1.5, 2.5], [0, 1, 2, 3], [0, 2, 4, 6])
        np.testing.assert_allclose(to_numpy(result), expected, rtol=1e-5)

    def test_masked_select(self):
        a = to_gpu(np.array([1, 2, 3, 4, 5], dtype=np.float64))
        mask = to_gpu(np.array([True, False, True, False, True], dtype=np.bool_))
        result = ins.masked_select(a, mask)
        assert result.numel == 3

    def test_where(self):
        cond = to_gpu(np.array([True, False, True], dtype=np.bool_))
        x = to_gpu(np.array([1, 2, 3], dtype=np.float64))
        y = to_gpu(np.array([4, 5, 6], dtype=np.float64))
        result = ins.where(cond, x, y)
        np.testing.assert_allclose(to_numpy(result), np.array([1, 5, 3]))

    def test_sort_2d(self):
        a = to_gpu(np.array([[3, 1], [2, 4]], dtype=np.float64))
        result = ins.sort(a)
        assert result.numel == 4

    def test_argsort_2d(self):
        a = to_gpu(np.array([[3, 1], [2, 4]], dtype=np.float64))
        result = ins.argsort(a)
        assert result.numel == 4

    def test_topk_smallest(self):
        a = to_gpu(np.array([1, 5, 3, 2, 4], dtype=np.float64))
        values, indices = ins.topk(a, 3, largest=False)
        data = sorted(to_numpy(values).tolist())
        assert data == [1.0, 2.0, 3.0]

    def test_scatter_reduce(self):
        a = to_gpu(np.zeros([4], dtype=np.float64))
        idx = to_gpu(np.array([0, 0, 2], dtype=np.int64))
        src = to_gpu(np.array([1, 2, 3], dtype=np.float64))
        result = ins.scatter_reduce(a, idx, src, "add", 0)
        assert result.numel == 4


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/cuda/julia/test_indexing.jl
# Indexing and sorting CUDA binding tests
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda julia tests/cuda/julia/test_indexing.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

# Try GPU
try
    Insight.load_backend("gpu")
    Insight.set_device(Insight.GPUPlace(0))
catch e
    println("SKIP: GPU backend not available: $e")
    exit(0)
end

passed = 0
failed = 0

function check(name, cond)
    global passed, failed
    if cond
        passed += 1
    else
        failed += 1
        println("FAIL: $name")
    end
end

println("=== Indexing CUDA ===")

# unique
a = Insight.to_array(Insight.from_data([3.0, 1.0, 2.0, 1.0, 3.0, 2.0]), Insight.GPUPlace(0))
result = Insight.unique_ins(a)
check("unique", Insight.numel(result.unique) >= 1)

# topk
a2 = Insight.to_array(Insight.from_data([1.0, 5.0, 3.0, 2.0, 4.0]), Insight.GPUPlace(0))
vals, idx = Insight.topk(a2, 3)
check("topk", Insight.numel(vals) == 3)

# topk smallest
vals2, idx2 = Insight.topk(a2, 3, largest=false)
check("topk_smallest", Insight.numel(vals2) == 3)

# gather
a3 = Insight.to_array(Insight.from_data([10.0, 20.0, 30.0, 40.0]), Insight.GPUPlace(0))
a3 = Insight.reshape(a3, [2, 2])
idx3 = Insight.to_array(Insight.from_data([0, 1, 1, 0]), Insight.GPUPlace(0))
idx3 = Insight.reshape(idx3, [2, 2])
g = Insight.gather(a3, 1, idx3)
check("gather", Insight.numel(g) == 4)

try
    Insight.gather(a3, 0, idx3)
    check("gather_axis0_invalid", false)
catch e
    check("gather_axis0_invalid", isa(e, ArgumentError))
end

# scatter
a4 = Insight.to_array(Insight.zeros(Int64[4], Insight.float64), Insight.GPUPlace(0))
idx4 = Insight.to_array(Insight.cast(Insight.from_data([0.0, 2.0]), Insight.int32), Insight.GPUPlace(0))
src = Insight.to_array(Insight.from_data([10.0, 20.0]), Insight.GPUPlace(0))
s = Insight.scatter(a4, 1, idx4, src)
check("scatter", Insight.numel(s) == 4)

# scatter_add
a5 = Insight.to_array(Insight.zeros(Int64[4], Insight.float64), Insight.GPUPlace(0))
idx5 = Insight.to_array(Insight.cast(Insight.from_data([0.0, 0.0, 2.0]), Insight.int32), Insight.GPUPlace(0))
src5 = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0]), Insight.GPUPlace(0))
sa = Insight.scatter_add(a5, 1, idx5, src5)
check("scatter_add", Insight.numel(sa) == 4)

# interp
xp = Insight.to_array(Insight.from_data([0.0, 1.0, 2.0, 3.0]), Insight.GPUPlace(0))
fp = Insight.to_array(Insight.from_data([0.0, 2.0, 4.0, 6.0]), Insight.GPUPlace(0))
x = Insight.to_array(Insight.from_data([0.5, 1.5, 2.5]), Insight.GPUPlace(0))
result = Insight.interp(x, xp, fp)
check("interp", Insight.numel(result) == 3)

# unique with counts
a6 = Insight.to_array(Insight.from_data([5.0, 3.0, 1.0, 3.0, 5.0]), Insight.GPUPlace(0))
result6 = Insight.unique_ins(a6, return_counts=true)
check("unique_counts", Insight.numel(result6.counts) >= 1)

# scatter_reduce
a7 = Insight.to_array(Insight.zeros(Int64[4], Insight.float64), Insight.GPUPlace(0))
idx7 = Insight.to_array(Insight.cast(Insight.from_data([0.0, 0.0, 2.0]), Insight.int32), Insight.GPUPlace(0))
src7 = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0]), Insight.GPUPlace(0))
sr = Insight.scatter_reduce(a7, 1, idx7, src7, reduce="add")
check("scatter_reduce", Insight.numel(sr) == 4)

# unique sorted
a8 = Insight.to_array(Insight.from_data([5.0, 3.0, 1.0, 3.0, 5.0]), Insight.GPUPlace(0))
result8 = Insight.unique_ins(a8)
check("unique_sorted", Insight.numel(result8.unique) >= 1)

# ============================================================================
# Results
# ============================================================================
println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0
    exit(1)
end
